r2/service.go

```diff
@@ -59,113 +78,569 @@ func NewService(iOpts instrument.Options, rootPrefix string, store Store) server
 	return &service{iOpts: iOpts, store: store, rootPrefix: rootPrefix}
 }
 
-func (s *service) readNamespaces(w http.ResponseWriter, _ *http.Request) {
-	w.WriteHeader(http.StatusNotImplemented)
-	fmt.Fprintf(w, "Placeholder for readNamespace")
+type payload struct {
+	data   []byte
+	status int
+}
+
+func newPayload(d []byte, status int) payload {
+	return payload{data: d, status: status}
+}
+
+func (p payload) write(w http.ResponseWriter) error {
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(p.status)
+	if _, err := w.Write(p.data); err != nil {
+		return fmt.Errorf("cannot write http payload. %s", err.Error())
+	}
+	return nil
+}
+
+// TODO(dgromov): Make this return a list of validation errors
+func parseRequest(s interface{}, body io.ReadCloser) string {
+	if err := json.NewDecoder(body).Decode(s); err != nil {
+		return fmt.Sprintf("Malformed Json: %s", err.Error())
+	}
+
+	// Invoking the validation explictely to have control over the format of the error output.
+	validate := validator.New()
+	validate.RegisterTagNameFunc(func(fld reflect.StructField) string {
+		parts := strings.SplitN(fld.Tag.Get("json"), ",", 2)
+		if len(parts) > 0 {
+			return parts[0]
+		}
+		return fld.Name
+	})
+
+	var required []string
+	if err := validate.Struct(s); err != nil {
+		for _, e := range err.(validator.ValidationErrors) {
+			if e.ActualTag() == "required" {
+				required = append(required, e.Namespace())
+			}
+		}
+	}
+
+	if len(required) > 0 {
+		return fmt.Sprintf("Required: [%v]", strings.Join(required, ", "))
+	}
+	return ""
+}
+
+func writeAPIResponse(w http.ResponseWriter, code int, msg string) error {
+	j, err := json.Marshal(apiResponse{Code: code, Message: msg})
+	if err != nil {
+		return err
+	}
+	if err := newPayload(j, code).write(w); err != nil {
+		return err
+	}
+	return nil
 }
 
-func (s *service) readNamespace(w http.ResponseWriter, _ *http.Request) {
-	w.WriteHeader(http.StatusNotImplemented)
-	fmt.Fprintf(w, "Placeholder for readNamespace")
+type r2Handler func(http.ResponseWriter, *http.Request) error
+
+func (fn r2Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if err := fn(w, r); err != nil {
+		handleError(w, err)
+	}
+}
+
+func handleError(w http.ResponseWriter, opError error) {
+	var err error
+	switch opError.(type) {
+	case ConflictError:
+		err = writeAPIResponse(w, http.StatusConflict, opError.Error())
+	case BadInputError:
+		err = writeAPIResponse(w, http.StatusBadRequest, opError.Error())
+	case VersionError:
+		err = writeAPIResponse(w, http.StatusConflict, opError.Error())
+	case NotFoundError:
+		err = writeAPIResponse(w, http.StatusNotFound, opError.Error())
+	default:
+		err = writeAPIResponse(w, http.StatusInternalServerError, opError.Error())
+	}
+
+	// Getting here means that the error handling failed. Trying to convey what was supposed to happen.
+	if err != nil {
+		http.Error(
```

Comment:
circular failure mode here ~ imagine if the http response writer was failing, your reaction here is to try again. maybe add a log of the error too